src/handlers/rendered_link.rs

```diff
@@ -25,32 +35,47 @@ pub async fn handle(ctx: &Context, event: &Event) -> anyhow::Result<()> {
 }
 
 async fn add_rendered_link(ctx: &Context, e: &IssuesEvent, prefix: &str) -> anyhow::Result<()> {
-    if e.action == IssuesAction::Opened {
+    if e.action == IssuesAction::Opened
+        || e.action == IssuesAction::Closed
+        || e.action == IssuesAction::Reopened
+    {
+        let edit = EditIssueBody::new(&e.issue, "RENDERED_LINK");
+        let current_data: Option<RenderedLinkData> = edit.current_data();
+
         let files = e.issue.files(&ctx.github).await?;
 
         if let Some(file) = files.iter().find(|f| f.filename.starts_with(prefix)) {
-            if !e.issue.body.contains("[Rendered]") {
-                // This URL should be stable while the PR is open, even if the
-                // user pushes new commits.
-                //
-                // It will go away if the user deletes their branch, or if
-                // they reset it (such as if they created a PR from master).
-                // That should usually only happen after the PR is closed.
-                // During the closing process, the closer should update the
-                // Rendered link to the new location (which we should
-                // automate!).
-                let head = e.issue.head.as_ref().unwrap();
-                let url = format!(
-                    "https://github.com/{}/blob/{}/{}",
-                    head.repo.full_name, head.git_ref, file.filename
-                );
-                e.issue
-                    .edit_body(
-                        &ctx.github,
-                        &format!("{}\n\n[Rendered]({})", e.issue.body, url),
-                    )
-                    .await?;
-            }
+            let mut current_data = current_data.unwrap_or_default();
+            let head = e.issue.head.as_ref().unwrap();
+
+            // This URL should be stable while the PR is open, even if the
+            // user pushes new commits.
+            //
+            // It will go away if the user deletes their branch, or if
+            // they reset it (such as if they created a PR from master).
+            // That should usually only happen after the PR is closed
+            // a which point we switch to a SHA-based url.
+            current_data.rendered_link = format!(
+                "https://github.com/{}/blob/{}/{}",
+                head.repo.full_name,
+                if e.action == IssuesAction::Closed {
+                    &head.sha
+                } else {
+                    &head.git_ref
+                },
+                file.filename
+            );
```

Comment:
I don't think this is going to be the correct link for what we want. I think for now the preference is:

If it is merged, link to `https://github.com/rust-lang/rfcs/blob/master/text/{filename}`. We don't want to link to the user's repo with a sha hash, because I don't think it will persist over time if deleted (the branch or the repo). And we probably want master to render any minor fixes made afterwards.

If it is not merged, link to the last known copy as `https://github.com/rust-lang/rfcs/blob/{sha}/text/{filename}`. I'm pretty sure those links should be permanent, as GitHub keeps a ref to the PR which should keep the blobs from being gc'd.

Reply:
Should we worry about missing some webhook that would make the "last known copy" (sha) no longer the "last known copy"?

I ask, because right now the link we produce is immunized against that.

Reply:
In the "If it is not merged" branch you used `{sha}` for the "last known copy", but since we only react to open, reopened, closed; using the SHA would not work as the PR is updated, we should be using `{git_ref}` or also react to push (and update the sha).

> I'm not too terribly concerned about that (and there's not much to be done about it anyways).

:+1:

Reply:
I've update the PR with the new logic. Might be easier to follow what I was saying.